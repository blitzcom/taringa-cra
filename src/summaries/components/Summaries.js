import _ from 'lodash'
import React, { Component } from 'react'
import { connect } from 'react-redux'

import infiniteScroll from '../../HOC/InfiniteScroll'
import Alert from '../../common/Alert'
import Summary from './Summary'
import * as actions from '../actions'
import { summariesSelector } from '../selectors'

export class Summaries extends Component {
  constructor (props) {
    super(props)
    this.state = {
      placeholderCount: props.placeholderCount || 8
    }
  }

  componentDidMount () {
    this.props.invalidate().then(() => this.props.loadMore())
  }

  renderPlaceholder () {
    return _.times(this.state.placeholderCount, i => <Summary key={i} isPlaceholder />)
  }

  render () {
    const { status, summaries } = this.props
    const isFetching = status === 'fetching'

    return (
      <div className="Summaries">
        { summaries.map(i => <Summary key={i.id} {...i} />) }

        {
          status === 'failure' && (
            <div className="my-4 text-center">
              <Alert className="my-3 font-weight-bold" type="danger">
                ¡Ratas! Parece que no estás conectado a internet.
                Intenta refrescar la página o pulsa el siguiente botón.
              </Alert>
              <button
                className="btn btn-outline-primary btn-sm"
                onClick={this.props.loadMore}
              >
                Volver a intentar
              </button>
            </div>
          )
        }

        {
          isFetching && this.renderPlaceholder()
        }
      </div>
    )
  }
}

Summaries.defaultProps = {
  loadMore: () => {},
  invalidate: () => Promise.resolve(),
  status: "success",
  summaries: [],
}

const checkForMoreContent = (state) => {
  const control = state.control.summariesFetch
  return control.ids.length < control.totalCount
}

const mapStateToProps = state => ({
  error: state.control.summariesFetch.error,
  hasMoreContent: checkForMoreContent(state),
  status: state.control.summariesFetch.status,
  summaries: summariesSelector(state),
})

const mapDispatchToProps = dispatch => ({
  loadMore: () => dispatch(actions.fetch()),
  invalidate: () => dispatch(actions.invalidate()),
})

export default connect(mapStateToProps, mapDispatchToProps)(
  infiniteScroll()(Summaries)
)
