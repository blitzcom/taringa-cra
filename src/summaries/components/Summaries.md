Initial Loading

```js
const Summaries = require('./Summaries').Summaries;
<Summaries status="fetching" />
```

Listing Summaries

```js
const Summaries = require('./Summaries').Summaries;
const shout = require('./data').StoryShout;
<Summaries stories={[shout]} />
```

Infinite loading

```js
const Summaries = require('./Summaries').Summaries;
const shout = require('./data').StoryShout;
<Summaries status="fetching" stories={[shout]} />
```

With error

```js

const Summaries = require('./Summaries').Summaries;
const shout = require('./data').StoryShout;
<Summaries
  status="failure"
  stories={[shout]}
  error="Network Error"
/>
```

With error and no stories

```js

const Summaries = require('./Summaries').Summaries;
<Summaries
  status="failure"
  error="Network Error"
/>
```
