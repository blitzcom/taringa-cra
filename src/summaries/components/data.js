export const SummaryShout = {
  id: "z50xj",
  created: 1513920502005,
  edited: null,
  title: "",
  nsfw: false,
  classic: {
    tid: "ctLMo",
    type: "shout"
  },
  tags: [
    "image",
    "classic"
  ],
  slug: "tienen-que-shoutear-minas-asi_z50xj",
  type: "story:summary",
  summary: {
    link: null,
    images: {
      slice: [
        {
          type: "image",
          url: "https://k60.kn3.net/taringa/1/8/3/5/9/3/15/f3d3_14/61D.jpg",
          width: 640,
          height: 774,
          format: "jpeg"
        }
      ],
      amount: 1
    },
    excerpt: "Tienen que shoutear minas asi",
    textLength: 29
  },
  channel: {
    type: "channel:summary",
    id: "1ld",
    created: "2016-08-05T03:09:30.000Z",
    name: "user-F3D3_14",
    title: "user-F3D3_14",
    channelType: "user_feed",
    thumbnail: "",
    background: "",
    description: "",
    category: "unknown/unknown",
    owner: {
      id: "1lc",
      created: "2016-08-05T03:09:30.000Z",
      type: "user",
      tid: 18359315,
      username: "F3D3_14",
      firstname: "Fedeformito el",
      lastname: "Perro CÃ³smico",
      birthday: "1998-1-8",
      avatar: "https://a06.t26.net/taringa/avatares/0/3/6/8/F/7/F3D3_14/120x120_885.jpg",
      site: "",
      gender: "m",
      country: "AR",
      message: "ðŸ˜œ Altoke Perro, te likeo hasta el corazÃ³n",
      migrated: [
        "classicCategories",
        "followers"
      ],
      subscribedToDefaultChannels: true
    },
    subscribers: 0,
    stories: 1753,
    latestStory: "2018-01-22T20:04:02.000Z",
    state: {
      subscription: "unsubscribed"
    }
  },
  owner: {
    id: "1lc",
    username: "F3D3_14",
    type: "user:summary",
    firstname: "Fedeformito el",
    lastname: "Perro CÃ³smico",
    avatar: "https://a06.t26.net/taringa/avatares/0/3/6/8/F/7/F3D3_14/120x120_885.jpg",
    gender: "m",
    country: "AR"
  },
  upvotes: 13799,
  downvotes: 100,
  shares: 4500,
  comments: 8000,
  state: {
    voted: 0
  },
  brandSafe: false,
  listViewSafe: false
}
