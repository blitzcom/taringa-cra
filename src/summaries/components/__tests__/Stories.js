import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import { Summaries } from '../Summaries'
import { StoryShout } from '../data'

describe('Summaries', () => {
  it('renders default', () => {
    const tree = renderer.create(<Summaries/>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders loader', () => {
    const tree = renderer.create(<Summaries status="fetching"/>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders error', () => {
    const tree = renderer
      .create(<Summaries status="failure" error="Network Error" />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders stories', () => {
    const tree = renderer
      .create(<Summaries stories={[StoryShout]} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('fetches stories on mount', () => {
    const mock = jest.fn()

    mount(<Summaries fetchStories={mock}/>)

    expect(mock).toBeCalled()
  })
})
