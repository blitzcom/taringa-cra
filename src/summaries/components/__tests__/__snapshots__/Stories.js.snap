// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summaries renders default 1`] = `
<div
  className="Summaries"
/>
`;

exports[`Summaries renders error 1`] = `
<div
  className="Summaries"
>
  <div
    className="my-4 text-center"
  >
    <div
      className="alert my-3 font-weight-bold alert-danger"
    >
      ¡Ratas! Parece que no estás conectado a internet. Intenta refrescar la página o pulsa el siguiente botón.
    </div>
    <button
      className="btn btn-outline-primary btn-sm"
      onClick={undefined}
    >
      Volver a intentar
    </button>
  </div>
</div>
`;

exports[`Summaries renders loader 1`] = `
<div
  className="Summaries"
>
  <div
    className="my-4 text-center"
  >
    <i
      className="fa fa-spin fa-spinner fa-2x"
    />
  </div>
</div>
`;

exports[`Summaries renders stories 1`] = `
<div
  className="Summaries"
/>
`;
